src/sparsezoo/objects/directories.py

```diff
@@ -283,14 +284,90 @@ class OnnxGz(Directory):
     available however, when the `path` property is accessed, it will point only
     to the `model.onnx` as this is the expected behavior for loading an onnx model
     with or without external data.
+
+    Class Invariants:
+        - `self.name` attribute of this class will point to the name of the onnx file
+        - `self._path` and `self.path` will point to the path of the extracted
+            onnx model
     """
 
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def __repr__(self):
+        return f"OnnxGz(name={self.name}, path={self._path})"
+
     @property
     def path(self):
-        _ = super().path  # call self.path to download initial file if not already
+        """
+        Assumptions:
+            - the tarball contains only one onnx model, with the
+                name `model.onnx`
+            - the tarball will be extracted to the same directory as the tarball
+
+        :post-condition: self._path will point to the path of the extracted
+            onnx model, and it exists
+        :return: path to the onnx model
+        """
+
+        expected_path: Path = (
+            Path(self._path)
+            if self._path is not None
+            else Path(self.parent_directory) / "model.onnx.tar.gz"
+        )
+
+        # _path can also point to parent directory of the tarball when
+        #  this object is initialized from files
+        if expected_path.is_dir():
+            # point expected to model.onnx.tar.gz
+            expected_path = expected_path / "model.onnx.tar.gz"
+
+        # point path to model.onnx.tar.gz before download/unzip
+        self._path = str(expected_path)
+
+        if not expected_path.exists():
+            # download the tarball if it does not exist
+            self.download()
+
+        self._check_if_extracted()
         if self.is_archive:
+            # if the tarball is not extracted, extract it
             self.unzip()
-        if os.path.isdir(self._path) and "model.onnx" in os.listdir(self._path):
-            # if unzipped into a directory, refer directly to model.onnx
-            self._path = os.path.join(self._path, "model.onnx")
+
+        onnx_model_path = expected_path.with_name(name="model.onnx")
```

Comment:
Do we need to enforce the model name as `model.onnx`? I think just searching for the file extension would be more flexible

Reply:
I agree it would be more flexible, but let's leave this as is for now as its the only check enforcing constraints on the contents of `model.onnx.tar.gz` we can update this if the constraints change?

Reply:
you could do something where you check if multiple *.onnx files are present. If not, ensure the single one present is `model.onnx`

Reply:
We would still need a definite plan on which of those onnx models should `Model(stub).onnx_model` point to; so I think this update would make more sense there?